### DISCUSSION TITLE
What is the best practice for constructing a Python DataFrame from data generated in C#?

### DISCUSSION THREAD

**Discussion by Jason5Lee** [pythonnet/pythonnet#2460] [Q&A]
I need to embed a C# library into Python. This C# library generates a data table, and my goal is to wrap it into a Python library that produces a DataFrame (e.g., `pandas.DataFrame`). Given that the data size can be large, I want to minimize overhead.
Currently, I plan to implement a C# method that generates the data and then calls Python to construct the DataFrame and then return it. This method will be called from Python, creating a "sandwich" structure where Python calls C#, which in turn calls back into Python, in the same Python engine. I can do this by passing delegates to C# that wraps the DataFrame creation function. However, I'm not sure that this is the most efficient approach. What are your thoughts?

**Comment by filmor** [comment]
What has worked for me quite well in the past for pure numerical data (before we had buffer API support which probably simplifies this!) was:

- Ask C# up-front about the expected dimensions
- Create a numpy array with those dimensions
- Get the raw pointer to the array using `ptr, read_only = arr.__array_interface__["data"]` (cf. https://numpy.org/doc/stable/reference/arrays.interface.html#object.__array_interface__)
- Pass this pointer to the real C# function that can then use [`Marshal.WriteInt64`](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.marshal.writeint64?view=net-8.0) or `Marshal.Copy` or something like that to write directly to the array

The more modern way available since 3.11 is to use the buffer protocol:
https://pythonnet.github.io/pythonnet/reference.html#function-Python.Runtime.PyObject.GetBuffer
